src/engines/julia/update_rules/nonlinear_conjugate.jl

```diff
@@ -0,0 +1,301 @@
+            haskey(interface_to_schedule_entry, inbound_interface) || error("The nonlinear node's backward rule uses the incoming message on the input edge to determine the approximation point. Try altering the variable order in the scheduler to first perform a forward pass.")
+            push!(inbounds, interface_to_schedule_entry[inbound_interface])
+        elseif node_interface == entry.interface
+            # Ignore inbound message on outbound interface
+            push!(inbounds, nothing)
+        elseif isClamped(inbound_interface)
+            # Hard-code outbound message of constant node in schedule
+            push!(inbounds, assembleClamp!(inbound_interface.node, Message))
+        else
+            # Collect message from previous result
+            push!(inbounds, interface_to_schedule_entry[inbound_interface])
+        end
+    end
+
+    # Push custom arguments if defined
+    if (node.dims !== nothing)
+        push!(inbounds, Dict{Symbol, Any}(:dims => node.dims[inx + 1],
+                                          :keyword => true))
+    end
+    if (node.n_samples !== nothing)
+        push!(inbounds, Dict{Symbol, Any}(:n_samples => node.n_samples,
+                                          :keyword   => true))
+    end
+    if (node.n_iterations !== nothing)
+        push!(inbounds, Dict{Symbol, Any}(:n_iterations => node.n_iterations,
+                                          :keyword      => true))
+    end
+    if (node.optimizer !== nothing)
+        push!(inbounds, Dict{Symbol, Any}(:optimizer => node.optimizer,
+                                          :keyword   => true))
+    end
+    return inbounds
+end
+
+
+#-------------------------
+# Optimization subroutines
+#-------------------------
+
+function renderCVI(log_μ_bw::Function,
+                   n_iterations::Int,
+                   optimizer::Any,
+                   λ_0::Vector,
+                   μ_fw::Message{F, V}) where {F<:FactorNode, V<:VariateType}
+
+    # Intialize natural parameters of forward message
+    η = naturalParams(μ_fw.dist)                   
+
+    # Initialize Fisher information matrix
+    A = λ -> logNormalizer(V, F, η=λ)
+    Fisher = λ -> ForwardDiff.hessian(A, λ)
+
+    # Initialize q marginal
+    λ_i = deepcopy(λ_0)
+    q_i = standardDistribution(V, F, η=λ_i)
+
+    for i=1:n_iterations
+        # Store previous results for possible reset
+        q_i_min = deepcopy(q_i)
+        λ_i_min = deepcopy(λ_i)
+
+        # Given the current sample, define natural gradient of q
+        s_q_i = sample(q_i)
+        log_q = λ -> logPdf(V, F, s_q_i, η=λ)
+        ∇log_q = λ -> ForwardDiff.gradient(log_q, λ)
+
+        # Compute current free energy gradient and update natural statistics
+        ∇log_μ_bw_i = log_μ_bw(s_q_i)*cholinv(Fisher(λ_i))*∇log_q(λ_i) # Natural gradient of backward message
+        ∇F_i = λ_i - η - ∇log_μ_bw_i # Natural gradient of free energy
+        λ_i -= apply!(optimizer, λ_i, ∇F_i) # Update λ_i
```

Comment:
I would use update! method instead of apply!, because I suspect that apply! might not update internal parameters that many optimizers have e.g. momentum. Following is from Flux's documentation "Flux internally calls on this function via the update! function. It shares the API with apply! but ensures that multiple parameters are handled gracefully."